langserver/config.go

```diff
@@ -26,30 +30,80 @@ import (
 type Config struct {
 	RPCTrace      string `yaml:"rpc_trace"`
 	PrometheusURL string `yaml:"prometheus_url"`
-	RESTAPIPort   int    `yaml:"rest_api_port"`
+	RESTAPIPort   uint64 `yaml:"rest_api_port"`
 }
 
-// ParseConfig parses a yaml configuration.
-//
-// It expects the content of the configuration file as its argument.
-func ParseConfig(in []byte) (*Config, error) {
-	var config Config
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	tmp := &Config{}
+	type plain Config
+	if err := unmarshal((*plain)(tmp)); err != nil {
+		return err
+	}
+	if err := tmp.Validate(); err != nil {
+		return err
+	}
+	*c = *tmp
+	return nil
+}
 
-	err := yaml.Unmarshal(in, &config)
+func (c *Config) UnmarshalENV() error {
+	prefix := "LANGSERVER"
+	conf := &struct {
+		RPCTrace      string
+		PrometheusURL string
+		// the envconfig lib is not able to convert an empty string to the value 0
+		// so we have to convert it manually
+		RESTAPIPort string
+	}{}
+	if err := envconfig.Process(prefix, conf); err != nil {
+		return err
+	}
+	if len(conf.RESTAPIPort) > 0 {
+		var parseError error
+		c.RESTAPIPort, parseError = strconv.ParseUint(conf.RESTAPIPort, 10, 64)
+		if parseError != nil {
+			return parseError
+		}
+	}
+	c.RPCTrace = conf.RPCTrace
+	c.PrometheusURL = conf.PrometheusURL
+	return c.Validate()
+}
 
-	return &config, err
+// Validate returns an error if the config is not valid
+func (c *Config) Validate() error {
+	if len(c.PrometheusURL) > 0 {
+		if _, err := url.Parse(c.PrometheusURL); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-// ParseConfigFile parses a yaml configuration file.
-//
-// It expects the path to a configuration file as its argument.
-func ParseConfigFile(path string) (*Config, error) {
-	data, err := ioutil.ReadFile(path)
+// ReadConfig gets the GlobalConfig from a configFile (that is a path to the file)
+func ReadConfig(configFile string) (*Config, error) {
+	if len(configFile) == 0 {
+		fmt.Fprintln(os.Stderr, "Config file empty, configuration is reading from System environment")
```

Comment:
This message is a bit misleading since having an empty config file is fine.
```suggestion
		fmt.Fprintln(os.Stderr, "No config file provided, configuration is reading from System environment")
```

Reply:
done